### PR TITLE
A0-1608 Add documentation of all public API and traits of aleph-client

### PR DESCRIPTION
**Pull request by Marcin-Radecki**
# Description

This PR adds documentation of all public API and traits of `aleph-client`, even if modules are not publicly exposed in the library itself. 

`#![warn(missing_docs)]` was added to `aleph-client's lib.rs and all (200+) warnings have been fixed. There are still some warnings `output filename collision.`, but they come from the Substrate itself, so can't be fixed by this PR.

Also there's some doubt about whether `aleph-client` should be part of the Rust workspace. This has some impact in the docs generation, ie `cargo doc` generates all deps doc by default, including Substrate packages, which takes a lot of time.. Not sure why `cargo doc --package aleph-client` does not limit that.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

# Checklist:

- I have created new documentation
- I have bumped aleph-client version if relevant

### CODE REVIEW

File: aleph-client/src/utility.rs [hunk 1]
```diff
@@ -6,17 +6,33 @@ use crate::{
...
+/// Any object that implements retrieving various info about blocks.
 #[async_trait::async_trait]
 pub trait BlocksApi {
+    /// Returns the first block of a session.
+    /// * `session` - number of the session to query the first block from
     async fn first_block_of_session(&self, session: SessionIndex) -> Option<BlockHash>;
+
+    /// Returns hash of a given block.
+    /// * `block` - number of the block
     async fn get_block_hash(&self, block: BlockNumber) -> Option<BlockHash>;
```
Comment (kostekIV) [inline aleph-client/src/utility.rs:18]: Returns block to smt like Returns hash of block if given block exists

File: aleph-client/src/utility.rs [hunk 2]
```diff
@@ -6,17 +6,33 @@ use crate::{
...
+
+    /// Returns the most recent block from the current best chain.
     async fn get_best_block(&self) -> BlockNumber;
+
+    /// Returns the most recent block from the finalized chain.
     async fn get_finalized_block_hash(&self) -> BlockHash;
+
+    /// Returns number of a given block hash.
```
Comment (kostekIV) [inline aleph-client/src/utility.rs:26]: same here `if block hash exists on the chain`

**Review by kostekIV — APPROVED** [review]
Astonishing job!

File: aleph-client/src/pallets/utility.rs [hunk 3]
```diff
@@ -1,7 +1,9 @@
 use crate::{api, BlockHash, Call, SignedConnection, TxStatus};
 
+/// Any object that implements pallet utility api.
```
Comment (fixxxedpoint) [inline aleph-client/src/pallets/utility.rs:3]: IMHO you shouldn't repeat everytime that a trait represents some set of objects, just focus on describing what that trait is doing.

**Review by fixxxedpoint — APPROVED** [review]
Very nice!

File: aleph-client/src/utility.rs (hunk 1 above)
Comment (Marcin-Radecki) [reply aleph-client/src/utility.rs]: done

File: aleph-client/src/utility.rs (hunk 2 above)
Comment (Marcin-Radecki) [reply aleph-client/src/utility.rs]: done

File: aleph-client/src/connections.rs [hunk 4]
```diff
@@ -90,6 +115,19 @@ impl Connection {
             .expect("There should be a value")
     }
 
+    /// Retrieves a decoded storage value stored under given storage key.
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/connections.rs]: done

File: aleph-client/src/connections.rs [hunk 5]
```diff
@@ -103,6 +141,21 @@ impl Connection {
...
 
+    /// Submit a RPC call.
+    ///
+    /// * `func_name` - name of a RPC call
+    /// * `params` - result of calling `rpc_params!` macro, that's `Vec<u8>` of encoded data
+    /// to this rpc call
+    ///
+    /// # Examples
+    /// ```rust
+    /// let func_name = "alephNode_emergencyFinalize";
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/connections.rs]: done

File: aleph-client/src/connections.rs [hunk 6]
```diff
@@ -116,14 +169,31 @@ impl Connection {
...
+    /// Creates new signed connection from existing [`Connection]` object.
+    /// * `connection` - existing connection
+    /// * `signer` - a [`KeyPair`] of signing account
     pub fn from_connection(connection: Connection, signer: KeyPair) -> Self {
         Self { connection, signer }
     }
-
+    /// Send a transaction to a chain. It waits for a given tx `status`.
+    /// * `tx` - encoded transaction payload
+    /// * `status` - tx status
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/connections.rs:191]: done

File: aleph-client/src/connections.rs [hunk 7]
```diff
@@ -162,10 +238,18 @@ impl SignedConnection {
 }
 
 impl RootConnection {
+    /// Creates new root connection from a given url.
+    /// By default, it tries to connect 10 times, waiting 1 second between each unsuccessful attempt.
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/connections.rs]: done

File: aleph-client/src/pallets/utility.rs (hunk 3 above)
Comment (Marcin-Radecki) [reply aleph-client/src/pallets/utility.rs]: done

File: aleph-client/src/connections.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/connections.rs]: sry, now done

**Pull request merged**